File: com.trollworks.gcs/src/com/trollworks/gcs/character/MiscPanel.java
/*
 * Copyright ©1998-2020 by Richard A. Wilkes. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, version 2.0. If a copy of the MPL was not distributed with
 * this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This Source Code Form is "Incompatible With Secondary Licenses", as
 * defined by the Mozilla Public License, version 2.0.
 */

package com.trollworks.gcs.character;

import com.trollworks.gcs.page.DropPanel;
import com.trollworks.gcs.page.PageField;
import com.trollworks.gcs.page.PageLabel;
import com.trollworks.gcs.ui.layout.PrecisionLayout;
import com.trollworks.gcs.ui.layout.PrecisionLayoutData;
import com.trollworks.gcs.utility.I18n;
import com.trollworks.gcs.utility.text.Text;

import javax.swing.SwingConstants;

/** The miscellaneous info panel. */
public class MiscPanel extends DropPanel {
    /**
     * Creates a new miscellaneous info panel.
     *
     * @param sheet The sheet to display the data for.
     */
    public MiscPanel(CharacterSheet sheet) {
        super(new PrecisionLayout().setColumns(2).setMargins(0).setSpacing(4, 0), I18n.Text("Miscellaneous"));
        createLabelAndDisabledField(sheet, GURPSCharacter.ID_CREATED, I18n.Text("Created:"), null);
        createLabelAndDisabledField(sheet, GURPSCharacter.ID_MODIFIED, I18n.Text("Modified:"), null);
        createLabelAndDisabledField(sheet, Settings.PREFIX, I18n.Text("Options:"), Text.wrapPlainTextForToolTip(I18n.Text("Each letter represents an optional rule. A uppercase letter indicates the rule is in use while a lowercase letter indicates the rule is not in use.")));
    }

    private void createLabelAndDisabledField(CharacterSheet sheet, String key, String title, String tooltip) {
        PageField field = new PageField(sheet, key, SwingConstants.LEFT, false, tooltip);
        add(new PageLabel(title, field), new PrecisionLayoutData().setEndHorizontalAlignment());
        add(field, new PrecisionLayoutData().setFillHorizontalAlignment().setGrabHorizontalSpace(true));
    }
}

File: ux/misc_panel.go
// Copyright (c) 1998-2026 by Richard A. Wilkes. All rights reserved.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, version 2.0. If a copy of the MPL was not distributed with
// this file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// This Source Code Form is "Incompatible With Secondary Licenses", as
// defined by the Mozilla Public License, version 2.0.

package ux

import (
	"github.com/richardwilkes/gcs/v5/model/colors"
	"github.com/richardwilkes/gcs/v5/model/gurps"
	"github.com/richardwilkes/gcs/v5/model/jio"
	"github.com/richardwilkes/toolbox/v2/i18n"
	"github.com/richardwilkes/unison"
)

// MiscPanel holds the contents of the miscellaneous block on the sheet.
type MiscPanel struct {
	unison.Panel
	entity    *gurps.Entity
	targetMgr *TargetMgr
	prefix    string
}

// NewMiscPanel creates a new miscellaneous panel.
func NewMiscPanel(entity *gurps.Entity, targetMgr *TargetMgr) *MiscPanel {
	m := &MiscPanel{
		entity:    entity,
		targetMgr: targetMgr,
		prefix:    targetMgr.NextPrefix(),
	}
	initTitledPagePanel(m, i18n.Text("Miscellaneous"), 2, true, colors.TintMisc)

	m.AddChild(NewPageLabelEnd(i18n.Text("Created")))
	m.AddChild(NewNonEditablePageFieldFor(func() string { return m.entity.CreatedOn.String() }))

	m.AddChild(NewPageLabelEnd(i18n.Text("Modified")))
	m.AddChild(NewNonEditablePageFieldFor(func() string { return m.entity.ModifiedOn.String() }))

	addLabeledStringPageField(m, m.targetMgr, m.prefix+"player", i18n.Text("Player"), NewPageLabelEnd,
		func() string { return m.entity.Profile.PlayerName },
		func(s string) { m.entity.Profile.PlayerName = s })

	return m
}

// UpdateModified updates the current modification timestamp.
func (m *MiscPanel) UpdateModified() {
	m.entity.ModifiedOn = jio.Now()
}

// SelectableTextField is a field whose text can be replaced wholesale and then selected. Both *unison.Field and the
// fields built on top of it, such as *NumericField, satisfy it. It exists so that SetTextAndMarkModified can be handed
// the outer field rather than the *unison.Field embedded within it, since the outer field may do work of its own when
// its text is replaced.
type SelectableTextField interface {
	unison.Paneler
	SetText(text string)
	SelectAll()
}

// SetTextAndMarkModified sets the field to the given text, selects it, requests focus, then calls MarkModified().
func SetTextAndMarkModified(field SelectableTextField, text string) {
	field.SetText(text)
	field.SelectAll()
	panel := field.AsPanel()
	panel.RequestFocus()
	panel.Parent().MarkForLayoutAndRedraw()
	MarkModified(field)
}
